File: packages/android/src/test/kotlin/dev/mushimushi/sdk/storage/OfflineQueueTest.kt
package dev.mushimushi.sdk.storage

import androidx.test.core.app.ApplicationProvider
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [33])
class OfflineQueueTest {
    @Test
    fun enqueuePeekClear() {
        val ctx = ApplicationProvider.getApplicationContext<android.content.Context>()
        val q = OfflineQueue(ctx, maxBytes = 100_000)
        q.enqueue(mapOf("description" to "hello", "category" to "bug"))
        q.enqueue(mapOf("description" to "world", "category" to "slow"))

        assertEquals(2, q.count())
        val peeked = q.peek(10)
        assertEquals(2, peeked.size)
        assertEquals("hello", peeked.first()["description"])

        q.clearDelivered(1)
        assertEquals(1, q.count())
    }

    @Test
    fun trimsOldestWhenOverBudget() {
        val ctx = ApplicationProvider.getApplicationContext<android.content.Context>()
        val q = OfflineQueue(ctx, maxBytes = 250)
        repeat(10) { i ->
            q.enqueue(mapOf("description" to "report-$i", "category" to "bug"))
        }
        // Each NDJSON line is ~44 B (Gson map + newline). Budget 250 B ⇒ five
        // newest rows survive after trimming the five oldest.
        assertEquals(5, q.count())
        assertEquals("report-5", q.peek(1).first()["description"])
    }
}

File: packages/flutter/test/offline_queue_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:mushi_mushi/src/offline_queue.dart';

void main() {
  late Directory tmp;
  late File file;

  setUp(() {
    tmp = Directory.systemTemp.createTempSync('mushi_test_');
    file = File('${tmp.path}/queue.ndjson');
  });

  tearDown(() {
    if (tmp.existsSync()) tmp.deleteSync(recursive: true);
  });

  test('enqueue, peek, clear', () async {
    final q = OfflineQueue(maxBytes: 100000, file: file);
    await q.enqueue({'description': 'hello', 'category': 'bug'});
    await q.enqueue({'description': 'world', 'category': 'slow'});

    expect(await q.count(), 2);
    final peeked = await q.peek(limit: 10);
    expect(peeked.length, 2);
    expect(peeked.first['description'], 'hello');

    await q.clearDelivered(1);
    expect(await q.count(), 1);
  });

  test('trims oldest when over byte budget', () async {
    final q = OfflineQueue(maxBytes: 250, file: file);
    for (var i = 0; i < 10; i++) {
      await q.enqueue({'description': 'report-$i', 'category': 'bug'});
    }
    // Matches iOS/Android contract: each ~44 B NDJSON line, 250 B cap
    // ⇒ five newest rows survive after trimming the five oldest.
    expect(await q.count(), 5);
    expect((await q.peek(limit: 1)).first['description'], 'report-5');
  });
}
